membrane_curvature/lib/mods.py

```diff
@@ -24,66 +38,58 @@ def grid_map(coords, factor):
     return index_grid_l, index_grid_m
 
 
-def core_fast_leaflet(universe, z_Ref, n_cells, selection, max_width):
+def derive_surface(n_cells, selection, max_width):
     """
-    Runs core_fast_leaflet for selected surface
+    Derive surface from distribution of z coordinates in grid.
 
     Parameters
     ----------
-    universe: MDA Universe
-        Provides coordinates and trajectory of the coordinates.
-    z_Ref : dict.
-        Dictionary to store values in every iteration over frames.
     n_cells : int.
         number of cells in the grid of size `max_width`.
     selection : AtomGroup
-        AtomGroup of reference selection to define the surface
+        AtomGroup of reference selected to define the surface
         of the membrane.
     max_width : int.
         Maximum width of simulation box.
 
     Returns
     -------
-
-    Returns a dictionary with tuples of [i,j] as keys, where 0<i<`max_width`,
-    and 0<j<`max_width`. The resulting values for each key are a list of
-    `len=(n_frames)` containing the extracted z_coordinate a given [i,j] cell
-    in each frame.
+    Returns set of z coordinates in grid.
 
     """
+    z_ref = np.zeros([n_cells, n_cells])
+    grid_z_coordinates = np.zeros([n_cells, n_cells])
+    grid_norm_unit = np.zeros([n_cells, n_cells])
 
-    grid_count_frames = np.zeros([n_cells, n_cells])
+    factor = np.float32(n_cells / max_width)
 
-    for ts in universe.trajectory:
+    for bead in selection:
```

Comment:
I don't think you need to (or should) loop over atoms.  Instead something like `np.rint(bead.positions[:, :2] / factor)` should give you an array of the `l` and `m` values.

Reply:
ANd you'll get rid of `get_positions()`.

Reply:
Thanks for your comment @richardjgowers. It also simplifies the function to the point `grid_map` isn't necessary. Since I also need the z coordinate as a float, I modified your suggestion to the following
`cell_xy_floor = np.int32( selection.positions[:,:2] / factor )`
`z_coordinate = selection.positions[:, 2] / factor * 10`.